lib/srv/desktop/rdp/rdpclient/src/lib.rs

```diff
@@ -591,12 +815,23 @@ pub unsafe extern "C" fn handle_tdp_sd_announce(
 ///
 /// # Safety
 ///
-/// The caller must ensure that res.fso.path points to a valid buffer.
+/// client_ptr MUST be a valid pointer.
+/// (validity defined by https://doc.rust-lang.org/nightly/core/primitive.pointer.html#method.as_ref-1)
+///
+/// res.fso.path MUST be a non-null pointer to a C-style null terminated string.
 #[no_mangle]
 pub unsafe extern "C" fn handle_tdp_sd_info_response(
     client_ptr: *mut Client,
     res: CGOSharedDirectoryInfoResponse,
 ) -> CGOErrCode {
+    // # Safety
+    //
+    // This function MUST NOT hang on to any of the pointers passed in to it after it returns.
+    // In other words, all pointer data that needs to persist after this function returns MUST
+    // be copied into Rust-owned memory.
+
+    let res = SharedDirectoryInfoResponse::from(res);
```

Comment:
At first, the `SharedDirectoryInfoResponse` was created in place where it was used, inside the call to `handle_tdp_sd_info_response`, but now it has been moved into top of the function. There's nothing wrong with defining it as a standalone variable, but I'm just wondering why you've moved it here? There are couple of old functions in which we follow the old pattern, so maybe we should unify them?

Reply:
The purpose of these changes are to always ensure that we adhere to the following safety protocol, all the way up here at the "C" boundary:

https://github.com/gravitational/teleport/blob/9f10dca018f7d0d195d9ebe21802680d5e5a7e4c/lib/srv/desktop/rdp/rdpclient/src/lib.rs#L838-L842

(Which we do in `SharedDirectoryInfoResponse::from`)

This way, we don't need to worry about any tricky memory bugs further down the call stack. It's all taken care of right here at the boundary.

I'm not sure which old functions you're referring to, can you point them out?

Reply:
Thanks for the response! I should've included the other function I had in my mind in my previous response. Like here: I did it the other way: https://github.com/gravitational/teleport/blob/9f10dca018f7d0d195d9ebe21802680d5e5a7e4c/lib/srv/desktop/rdp/rdpclient/src/lib.rs#L984-L1004

Reply:
Ah I see, good eye. I must have not had those functions merged yet whenever I went back and revised all of these functions to use this pattern. The code and memory all works fine, but I'll add the safety comments and rearrange the order in an upcoming PR just for the sake of consistency.